rosbag2_compression/src/rosbag2_compression/zstd_compressor.cpp

```diff
@@ -0,0 +1,120 @@
+  const std::string & uri,
+  size_t & decompressed_buffer_length)
+{
+  try {
+    std::ifstream infile{uri};
+    infile.exceptions(std::ifstream::failbit);
+    // Get size and allocate
+    infile.seekg(0, std::ios::end);
+    decompressed_buffer_length = infile.tellg();
+    auto decompressed_buffer = std::make_unique<char[]>(decompressed_buffer_length);
+    // Go back and read in contents
+    infile.seekg(0 /* off */, std::ios::beg);
+    infile.read(decompressed_buffer.get(), decompressed_buffer_length);
+    return decompressed_buffer;
+  } catch (std::ios_base::failure & fail) {
+    ROSBAG2_COMPRESSION_LOG_ERROR_STREAM(
+      "Caught IO stream exception while reading file: " << fail.what());
+    throw fail;
+  }
+}
+
+void check_compression_result(const size_t compression_result)
+{
+  if (ZSTD_isError(compression_result)) {
+    std::stringstream error;
+    error << "ZSTD compression error: " << ZSTD_getErrorName(compression_result);
+    throw std::runtime_error(error.str());
+  }
+  ROSBAG2_COMPRESSION_LOG_DEBUG("ZSTD compressed file.");
+}
+
+void print_compression_statistics(
+  std::chrono::high_resolution_clock::time_point start,
+  std::chrono::high_resolution_clock::time_point end,
+  size_t decompressed_size, size_t compressed_size)
+{
+  const auto duration = std::chrono::duration_cast<std::chrono::microseconds>(end - start);
+  const auto compression_ratio =
+    static_cast<double>(decompressed_size) / static_cast<double>(compressed_size);
+  ROSBAG2_COMPRESSION_LOG_DEBUG_STREAM(
+    "Compression statistics:\n" <<
+      "Time: " << duration.count() << " microseconds" <<
+      "Compression Ratio: " << compression_ratio
+  );
+}
+}  // namespace
+
+namespace rosbag2_compression
+{
+
+std::string ZstdCompressor::compress_uri(const std::string & uri)
+{
+  const auto start = std::chrono::high_resolution_clock::now();
+  const auto compressed_uri = uri + "." + get_compression_identifier();
+  try {
+    size_t decompressed_buffer_length{0};
+    const auto decompressed_buffer = get_input_buffer(uri, decompressed_buffer_length);
+    // Allocate based on compression bound and compress
+    const size_t compressed_buffer_length = ZSTD_compressBound(decompressed_buffer_length);
+    auto compressed_buffer = std::make_unique<char[]>(compressed_buffer_length);
+    // Perform compression and check
+    const size_t compression_result = ZSTD_compress(
+      compressed_buffer.get(), compressed_buffer_length,
+      decompressed_buffer.get(), decompressed_buffer_length, DEFAULT_ZSTD_COMPRESSION_LEVEL);
+    check_compression_result(compression_result);
+    std::ofstream outfile{compressed_uri, std::ios::out | std::ios::binary};
+    outfile.write(compressed_buffer.get(), compression_result);
+    const auto end = std::chrono::high_resolution_clock::now();
+    print_compression_statistics(start, end, decompressed_buffer_length, compression_result);
+  } catch (std::ios_base::failure & fail) {
```

Comment:
I think this try/catch is covering way too much. I think the try/catch should cover as little as possible and thus only the catch the related exception. What call to `ifstream` can actually really fail here?

Reply:
Agreed, it can probably be down sized to just when writing out the file.